### PR TITLE
[REG-1872] add validation rules + update validations in general

### PR DESCRIPTION
**Pull request by addisonbgross**
# This should not be merged until we're ready to release the Validation Rules feature

## What has been done
- Added reference docs about Validation Rules
- Updated the existing Validation docs, not a tonne of stuff

### Validations

image
image
image
image
image
image

### Rules Reference
image

### CODE REVIEW

File: docs/validations/validations-rules-reference.mdx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+---
+sidebar_label: 'Rules Reference'
+---
+
+# Rules Reference
+
+Rules can be asserted against a gameplay recording to ensure that the recording meets specific criteria. We provide a set of rule types that can be used to construct assertions based on various performance metrics.
```
Comment (abeizer) [inline docs/validations/validations-rules-reference.mdx:7]: I think we're missing an answer to the question "How is this different than a validation scenario?". Probably want to reword this to include that rules are applied to every tick in the gameplay session and act more as monitoring than a validation.

We'll also want to clarify somewhere that rules produce warnings, not errors.

File: docs/validations/validations-getting-started.mdx [hunk 2]
```diff
@@ -6,37 +6,49 @@ sidebar_label: 'Getting Started with Validations'
 
 ![Screenshot of a scenario](./img/scenario-example.png)
 
-**Validations** are used to assert states within you game. You can create the scenarios you want to verify using our Scenario Builder,
-and can run collections of scenarios in Unity.
+**Validations** are used to assert states within you game. You can create the Scenarios you want to verify using our Scenario Builder,
```
Comment (abeizer) [inline docs/validations/validations-getting-started.mdx:9]:
```suggestion
**Validations** are used to assert states within your game. You can create Validation Scenarios using our Scenario Builder,
```

File: docs/validations/validations-getting-started.mdx [hunk 3]
```diff
@@ -6,37 +6,49 @@ sidebar_label: 'Getting Started with Validations'
...
+and can run collections of Scenarios in Unity.
```
Comment (abeizer) [inline docs/validations/validations-getting-started.mdx:10]: This currently isn't true. Probably want to remove this bit for now and we can provide a walkthrough from the Unity-side if/when that's in place

File: docs/validations/validations-getting-started.mdx [hunk 4]
```diff
@@ -6,37 +6,49 @@ sidebar_label: 'Getting Started with Validations'
...
+Rules can be added to a Suite by clicking the "Create Rule" button on the Suite's page. You can then select the Rule type you want to add, and configure the Rule's parameters. After adding a Rule to a Suite, you can always edit it later.
+
+![Creating a Validation Rule](./img/create-validation-rule.png)
+
+Check out the [full Rules reference](./validations-rules-reference) for more details
+
 ## Running Validations
 
-You can either run **all** of the Scenarios within a Suite, or an individual Scenario. To run a Suite, click the "Run Suite" button, and select a Gameplay Session to test against. When the Suite is finished running, you can view the results of each Scenario within the Suite on the Test Runs page.
+You can either run **all** of the Scenarios within a Suite, or test run an individual Scenario. To run a Suite, click the "Run Suite" button, and select a Gameplay Session to test against. When the Suite is finished running, you can view the results of each Scenario and Rule within the Suite on the Test Runs page.
```
Comment (abeizer) [inline docs/validations/validations-getting-started.mdx:44]: "test run" sounds awkward in this context. Does this sound any better?

```suggestion
You can either run **all** of the Scenarios within a Suite, or trial an individual Scenario. To run a Suite, click the "Run Suite" button, and select a Gameplay Session to test against. When the Suite is finished running, you can view the results of each Scenario and Rule within the Suite on the Test Runs page.
```

File: docs/validations/validations-getting-started.mdx [hunk 5]
```diff
@@ -6,37 +6,49 @@ sidebar_label: 'Getting Started with Validations'
...
 
-To run an individual Scenario, click the "Run Scenario" button within the Scenario's menu. The Scenario's results will be shown in a dialog.
+To test run an individual Scenario, click the "Run Scenario" button within the Scenario's menu. The Scenario's results will be shown in a dialog, and will not be persisted.
 
 ![Screenshot of a scenario](./img/run-scenario.png)
 
 ## Viewing Results
 
-To access the results of your Suite or Scenario, navigate to the Test Runs page. Here you can see the results of each Scenario within a Suite. You can also view the results of a Scenario's steps, and the state of the game at each tick during the Scenario.
+To access the results of your Suite, navigate to the Test Runs page. Here you can see the results of each Scenario  and Rule within a Suite. You can also view the results of a Scenario's steps, and the state of the game at each tick during the Scenario.
```
Comment (abeizer) [inline docs/validations/validations-getting-started.mdx:52]:
```suggestion
To access the results of your Suite, navigate to the Test Runs page. Here you can see the results of each Scenario and Rule within a Suite. You can also view the results of a Scenario's steps, and the state of the game at each tick during the Scenario.
```

File: docs/validations/validations-rules-reference.mdx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+Rules belong to a Validation Suite, and are asserted each time the Validation Suite is run.
+
+## Rule Types
+
+### Frames Per Second (with percentile)
+
+Assert that the frames per second of the gameplay recording are above a threshold, within a specified percentile.
```
Comment (abeizer) [inline docs/validations/validations-rules-reference.mdx:15]: These descriptions need to account for the fact that meeting the threshold does not violate it. 

"are above a threshold" -> "do not fall below the threshold"
similar for the other rule descriptions

**Review by abeizer — CHANGES_REQUESTED** [review]

File: docs/validations/validations-rules-reference.mdx [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+### Average Frames Per Second
+
+Assert that the average frames per second of the gameplay recording is above a threshold.
```
Comment (svolokh) [inline docs/validations/validations-rules-reference.mdx:19]: I think we need to clarify how this average is calculated. i.e. whether this is an average over the entire gameplay recording or whether this is a rolling average over the last N ticks/frames.

File: docs/validations/validations-rules-reference.mdx [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+### Maximum Frame Time
+
+Assert that the maximum frame time of the gameplay recording is below a threshold.
```
Comment (svolokh) [inline docs/validations/validations-rules-reference.mdx:23]: Readers will likely be confused about what frame time is. We should clarify that this represents the time spent calculating and rendering a frame and is the inverse of FPS.

File: docs/validations/validations-rules-reference.mdx (hunk 6 above)
Comment (addisonbgross) [reply docs/validations/validations-rules-reference.mdx:15]: I agree. I have the 3 descriptions containing either "above a threshold", or "below a threshold"

File: docs/validations/validations-rules-reference.mdx (hunk 1 above)
Comment (addisonbgross) [reply docs/validations/validations-rules-reference.mdx]: good call, added a bit about this

File: docs/validations/validations-rules-reference.mdx (hunk 7 above)
Comment (addisonbgross) [reply docs/validations/validations-rules-reference.mdx]: thanks. Added

File: docs/validations/validations-rules-reference.mdx (hunk 8 above)
Comment (addisonbgross) [reply docs/validations/validations-rules-reference.mdx]: good idea added

**Review by vontell — APPROVED** [review]

File: docs/validations/validations-rules-reference.mdx (hunk 6 above)
Comment (abeizer) [reply docs/validations/validations-rules-reference.mdx:13]:
```suggestion
Assert that the frames per second of the gameplay recording do not fall below a threshold, within a specified percentile.
```

File: docs/validations/validations-rules-reference.mdx [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+## Rule Types
+
+### Frames Per Second (with percentile)
+
+Assert that the frames per second of the gameplay recording are above a threshold, within a specified percentile.
+
+### Average Frames Per Second
+
+Assert that the average frames per second of the gameplay recording is above a threshold. This average is computed
```
Comment (abeizer) [inline docs/validations/validations-rules-reference.mdx:17]:
```suggestion
Assert that the average frames per second of the gameplay recording does not fall below a threshold. This average is computed
```

File: docs/validations/validations-rules-reference.mdx [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+across all frames in a gameplay recording.
+
+### Maximum Frame Time
+
+Assert that the maximum frame time of the gameplay recording is below a threshold. Frame time is the amount of time spent calculating and rendering a frame, and is measured in milliseconds. If a single frame exceeds the specified threshold, the rule will be violated.
```
Comment (abeizer) [inline docs/validations/validations-rules-reference.mdx:22]:
```suggestion
Assert that the maximum frame time of the gameplay recording does not exceed a threshold. Frame time is the amount of time spent calculating and rendering a frame, and is measured in milliseconds. If a single frame exceeds the specified threshold, the rule will be violated.
```

**Review by abeizer — CHANGES_REQUESTED** [review]

**Review by svolokh — APPROVED** [review]

File: docs/validations/validations-rules-reference.mdx (hunk 6 above)
Comment (addisonbgross) [reply docs/validations/validations-rules-reference.mdx]: ah there was some confusion on my part regarding your original comment. Changes made

Comment (addisonbgross) [reply docs/validations/validations-rules-reference.mdx]: f
